### PR TITLE
add GROBID author extraction from PDF to editor

### PR DESCRIPTION
**Pull request by nooraangelva**
ref: https://github.com/cern-sis/issues-inspire/issues/49

### CODE REVIEW

File: backend/tests/integration/editor/test_editor_views.py [hunk 1]
```diff
@@ -951,3 +951,72 @@ def test_refextract_text_dedupe_references(mock_refs, inspire_app):
...
+            "https://arxiv.org/pdf/1612.06414.pdf",
+            content=pkg_resources.resource_string(
+                __name__, os.path.join("fixtures", "1612.06414.pdf")
+            ),
+        )
+        requests_mocker.register_uri(
+            "GET",
+            "http://test-indexer:9200/records-hep/hep/_search?_source=control_number",
+            json=es_response,
+        )
```
Comment (MJedr) [inline backend/tests/integration/editor/test_editor_views.py:979]: can't we do it with vcr instead? I would say it is very `next` way of writing tests :p

File: backend/inspirehep/editor/views.py [hunk 2]
```diff
@@ -379,3 +369,61 @@ def remove_editor_lock(endpoint, pid_value):
...
+    data = {"input": document}
+    data.update(kwargs)
+    grobid_url = current_app.config["GROBID_URL"]
+    response = requests.post(urljoin(grobid_url, grobid_api_path), files=data)
+    response.raise_for_status()
+
+    # Clean GrobidAuthors response
+    parsed_authors = []
+    for author in GrobidAuthors(response.text).parse_all():
+        author.pop("parsed_affiliations")
```
Comment (MJedr) [inline backend/inspirehep/editor/views.py:405]: I don't get it why you need to pop it if you add explicitly `author` key later and you are not using `author` var

File: backend/inspirehep/editor/views.py [hunk 3]
```diff
@@ -379,3 +369,61 @@ def remove_editor_lock(endpoint, pid_value):
...
+        parsed_authors.append(author["author"])
+
+    if not response:
+        return
+    return {"authors": parsed_authors}
+
+
+def normalize_affiliations_for_authors(parsed_authors):
```
Comment (MJedr) [inline backend/inspirehep/editor/views.py:413]: could you use this function in the other endpoint that is normalising affiliations (authorlist afair)?

File: backend/inspirehep/submissions/grobid_authors_parsers.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
+# -*- coding: utf-8 -*-
```
Comment (MJedr) [inline backend/inspirehep/submissions/grobid_authors_parsers.py:1]: I don't know if submissions is the place for that, maybe place it in editor/utils?

File: backend/inspirehep/editor/views.py (hunk 3 above)
Comment (nooraangelva) [reply backend/inspirehep/editor/views.py:413]: I'm using it there already 👍

File: backend/inspirehep/editor/authorlist_utils.py [hunk 5]
```diff
@@ -1,7 +1,10 @@
+from __future__ import absolute_import, division, print_function
```
Comment (MJedr) [inline backend/inspirehep/editor/authorlist_utils.py:1]: this is for python2 (it is from next), shouldn't be needed

**Review by MJedr — COMMENTED** [review]
just a small comment

File: backend/inspirehep/editor/config.py [hunk 6]
```diff
@@ -6,3 +6,4 @@
 # the terms of the MIT License; see LICENSE file for more details.
 
 EDITOR_UPLOAD_ALLOWED_EXTENSIONS = {".pdf"}
+GROBID_URL = "https://grobid.inspirebeta.net"
```
Comment (MJedr) [inline backend/inspirehep/editor/config.py:9]: could you change it to a dummy url maybe?

**Review by MJedr — COMMENTED** [review]
looks good, did you test editor part @nooraangelva?

**Comment by nooraangelva** [comment]
> looks good, did you test editor part @nooraangelva?
Yep, I tested it locally yesterday.

File: backend/inspirehep/editor/authorlist_utils.py [hunk 7]
```diff
@@ -405,3 +406,128 @@ def authorlist(text):
             )
         result["authors"] = builder.record["authors"]
     return result
+
+
+class GrobidAuthors(object):
+    def __init__(self, xml_text):
+
```
Comment (drjova) [inline backend/inspirehep/editor/authorlist_utils.py:413]: could you please remove the empty line?

File: backend/inspirehep/editor/authorlist_utils.py [hunk 8]
```diff
@@ -405,3 +406,128 @@ def authorlist(text):
             )
         result["authors"] = builder.record["authors"]
     return result
+
+
+class GrobidAuthors(object):
```
Comment (drjova) [inline backend/inspirehep/editor/authorlist_utils.py:411]: I would move these two classes to https://github.com/inspirehep/inspirehep/blob/master/backend/inspirehep/matcher/parsers.py

File: backend/inspirehep/editor/views.py [hunk 9]
```diff
@@ -379,3 +368,58 @@ def remove_editor_lock(endpoint, pid_value):
...
+@login_required
+def authorlist_url():
+    """GROBID extraction from PDF"""
+
+    url = request.json["url"]
+    api_path = "api/processHeaderDocument"
+    kwargs_to_grobid = {"includeRawAffiliations": "1", "consolidateHeader": "1"}
+
+    try:
+
```
Comment (drjova) [inline backend/inspirehep/editor/views.py:383]: no need of the extra space :)

File: backend/inspirehep/editor/views.py [hunk 10]
```diff
@@ -379,3 +368,58 @@ def remove_editor_lock(endpoint, pid_value):
...
+        parsed_authors = post_pdf_to_grobid(url, api_path, **kwargs_to_grobid)
+        authors_normalized_affs = normalize_affiliations_for_authors(parsed_authors)
+        return jsonify(authors_normalized_affs)
+
```
Comment (drjova) [inline backend/inspirehep/editor/views.py:387]: no need of the extra space :)

File: backend/inspirehep/editor/views.py [hunk 11]
```diff
@@ -379,3 +368,58 @@ def remove_editor_lock(endpoint, pid_value):
...
+    grobid_url = current_app.config["GROBID_URL"]
+    response = requests.post(urljoin(grobid_url, grobid_api_path), files=data)
+    response.raise_for_status()
+
+    parsed_authors = [
+        author["author"] for author in GrobidAuthors(response.text).parse_all()
+    ]
+    if not response:
+        return
+    return {"authors": parsed_authors}
```
Comment (drjova) [inline backend/inspirehep/editor/views.py:406]: We do the following:
* move this function to https://github.com/inspirehep/inspirehep/blob/master/backend/inspirehep/matcher/api.py 
* name it `get_affiliations_from_pdf`
* no need to pass the `grobid_api_path` param

File: backend/inspirehep/editor/views.py [hunk 12]
```diff
@@ -379,3 +368,58 @@ def remove_editor_lock(endpoint, pid_value):
     )
     editor_soft_lock.remove_lock()
     return jsonify(success=True)
+
+
+@blueprint.route("/authorlist/url", methods=["POST"])
+@login_required
```
Comment (drjova) [inline backend/inspirehep/editor/views.py:374]: this login will allow any kind of user, we should use `@login_required_with_roles([Roles.cataloger.value])`

File: backend/inspirehep/matcher/api.py [hunk 13]
```diff
@@ -11,6 +11,9 @@
 from inspire_matcher import match
 from inspire_utils.dedupers import dedupe_list
 from inspire_utils.record import get_value
+from six.moves.urllib.parse import urljoin
```
Comment (MJedr) [inline backend/inspirehep/matcher/api.py:14]: can you import from `urllib.parse import urljoin`?

File: backend/inspirehep/matcher/parsers.py [hunk 14]
```diff
@@ -83,3 +83,127 @@ def year(self):
         return self.root.xpath(
             "/biblStruct/monogr/imprint/date[@type='published']/@when"
         ).get()
+
+
+class GrobidAuthors(object):
+    def __init__(self, xml_text):
+        self._xml = Selector(text=xml_text, type="xml")
+
```
Comment (MJedr) [inline backend/inspirehep/matcher/parsers.py:91]: no need for an extra line

**Review by MJedr — APPROVED** [review]
let's test it! 👍

**Pull request merged**